### ISSUE TITLE
useContainerQuery hook proposal 

### ISSUE THREAD

**Issue by mayank23** [react-container-query/react-container-query#92]
Hi 😄 , we've been using react-container-query in production for a while. However with the latest React releases, legacy lifecycle methods and apis such as `findDOMNode` will be deprecated, and warnings already are emitted by React.

I think it would be helpful to export a `useContainerQuery` hook. I wanted to gather some feedback first before making a full PR, so here's an initial implementation of the `useContainerQuery` hook that I experimented with in my fork and have working. (Right now I have it exported from `src/index.tsx`):

**Example Code Sandbox that uses the hook: https://codesandbox.io/s/zen-violet-9kbm3**

My Fork: https://github.com/react-container-query/react-container-query/compare/master...mayank23:use-container-query-hook?expand=1

## Hook Implementation

> I ported the following class component to be a hook: https://github.com/react-container-query/react-container-query/blob/master/src/index.tsx#L16-L63

```javascript
export const useContainerQuery = (query: Query, initialSize: Size) => {
  const [params, setParams] = React.useState(() => {
    if (!initialSize) {
      return {};
    }
    return matchQueries(query)(initialSize);
  });

  const [containerRef, setContainerRef] = React.useState(null);

  const refCallback = React.useCallback((node) => {
    setContainerRef(node);
  }, []);
  
  React.useEffect(() => {
    if (containerRef) {
      let cqCore: ContainerQueryCore | null = new ContainerQueryCore(
        query,
        (params) => {
          setParams(params);
        }
      );
      cqCore.observe(containerRef);
      return () => {
        cqCore!.disconnect();
        cqCore = null;
      };
    }
  }, [query, containerRef]);

  return [params, refCallback];
};
```

### Section 1

```javascript
  const [params, setParams] = React.useState(() => {
    if (!initialSize) {
      return {};
    }
    return matchQueries(query)(initialSize);
  });
```
- `params` would hold the matched class names that ContainerQueryCore will set based on the size of the container element. 
- if `initialSize` is passed, then when the hook first renders the initial state for `params` can be computed based on the query and initial size. 

#### Section 2
```javascript
  const [containerRef, setContainerRef] = React.useState(null);
  const refCallback = React.useCallback((node) => {
    setContainerRef(node);
  }, []);
```
- instead of `findDOMNode` that is currently used, we store a reference to the container element using a combination of `state` and a `callback ref`. For more details see the following: https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node and  https://reactjs.org/docs/refs-and-the-dom.html#callback-refs
- We use `React.useState` to hold a reference to the container element 
- After mounts the container element to the DOM, our `refCallback` function will be invoked
- We would save the reference to the desired container element DOM node with `setContainerRef`
- This in turn would cause a re-render of the `useContainerQuery` hook, and would trigger the effect code below.
- More Details on a user would set the `ref` is described in [Example 1](#example-1-queries-against-a-native-dom-element).

#### Section 3
```javascript
  React.useEffect(() => {
    if (containerRef) {
      let cqCore: ContainerQueryCore | null = new ContainerQueryCore(
        query,
        (params) => {
          setParams(params);
        }
      );
      cqCore.observe(containerRef);
      return () => {
        cqCore!.disconnect();
        cqCore = null;
      };
    }
  }, [query, containerRef]);
```
  - this effect hook will run on mount and whenever the `query` or `containerRef` (container element reference) changes.
  - when the effect is ran, it'll check if we have a reference to a container element before doing anything
  - if we have a reference to a container element, it'll instantiate the `ContainerQueryCore` using the current query object and set it to observe the container element (`containerRef`)
  - `ContainerQueryCore` would detect the size of the element and then match the queries. Once the matching is complete, our callback passed to it would set the latest params state: `setParams(params)`. This would trigger a re-render of the component using the hook with the latest `params`.
  - whenever the query or container element is changed or when the component that uses this `useContainerQuery` hook unmounts, the returned effect cleanup function will before the subsequent effect with a new query or container element runs or do a final cleanup.
    - ```javascript
      return () => {
        cqCore!.disconnect();
        cqCore = null;
      };
      ```

#### Section 4
```javascript
  return [params, refCallback];
```
  - the `useContainerQuery` hook returns the `params` (matched classnames object based on the current query) and the `refCallback` function that should be set on the desired container element that the user wants to match container queries on. e.) `<div ref={refCallback}></div>`

### Example 1 - Queries against a native DOM element

> Compare the hook style code with the original example from https://github.com/react-container-query/react-container-query#containerquery-queryquery-initialsizewidth-height 

```jsx
import React from 'react';
import { useContainerQuery } from 'react-container-query';

const query = {
  'width-between-400-and-599': {
    minWidth: 400,
    maxWidth: 599,
  },
  'width-larger-than-600': {
    minWidth: 600,
  },
};

const MyComponent = () => {
  const [params, containerRef] = useContainerQuery(query);
  return <div ref={containerRef} className={classnames(params)}>the box</div>;
};
```
- the end user sets the `containerRef` ref callback function, that the `useContainerQuery` hook returns, on the element they want to run container queries against. In this case we set it on the `div` element which wraps `the box` text.
- when the component mounts, react will invoke the `containerRef` callback function with the actual DOM node reference of the `div` element. 
- the `containerRef` callback function would in turn call `setContainerRef(divNode)` within the hook , and this would save node reference as state. 
- The state update would cause the `useContainerQuery` hook to re-render (and also `MyComponent`)  and the `useEffect` hook would run due to a new `containerRef` being set
- as a result, a new `ContainerQueryCore` instance would be created to observe this element.
- `ContainerQueryCore` would match the queries based on the size of the `div` and set the state for the latest `params`.
- Now `MyComponent` would re-render with the latest `params`. 
  ```javascript
  {
    'width-between-400-and-599': true,
    'width-larger-than-600': false
  }
  ```

### Example 2 - Usage for a non-native container element - React.forwardRef

If the container element we want to set the `containerRef` on is React component and not a native `DOM` element such as `div`, `span`, etc. Then the developer can leverage `React.forwardRef`. 

```jsx
import React from 'react';
import { useContainerQuery } from 'react-container-query';

const query = {
  'width-between-400-and-599': {
    minWidth: 400,
    maxWidth: 599,
  },
  'width-larger-than-600': {
    minWidth: 600,
  },
};

const MyCustomWrapper = React.forwardRef((props, ref) => {
  return <div ref={ref}>{props.children}</div>
});

const MyComponent = () => {
  const [params, containerRef] = useContainerQuery(query);
  return <MyCustomWrapper ref={containerRef} className={classnames(params)}>the box</div>;
};
```

- In this case, `<MyCustomWrapper />` would forward the `containerRef` and set it on the `div` react element it is using to wrap the children.

Would appreciate any feedback/suggestions for the hook and how we can get this merged into master if that is what the maintainers would like to do. 

pings: @pixelbandito @d6u 

Thanks!

**Comment by pixelbandito** [comment]
I'm going to have to look into this a little more after work, but my first impression is ❤️ !

Feel free to put up a PR. It could also use tests and a README update.

**Comment by jfnault-seedbox** [comment]
This would be great!

**Comment by mayank23** [comment]
PR: https://github.com/react-container-query/react-container-query/pull/93

**Comment by mayank23** [comment]
Going to close, thanks for all the help with this @pixelbandito !
